### PR TITLE
Fix arm build

### PR DESCRIPTION
**Pull request by mikemhenry**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by mikemhenry** [comment]
Progress! Now hitting:
```
/bin/sh: rule_parse: command not found
make[2]: *** [AmberTools/src/nab/CMakeFiles/nab_checkexpr_obj.dir/build.make:77: AmberTools/src/nab/checkexpr.c] Error 127
make[1]: *** [CMakeFiles/Makefile2:3713: AmberTools/src/nab/CMakeFiles/nab_checkexpr_obj.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

It looks like `rule_parse` is a small program that ambertools needs but it should build it

**Comment by mikemhenry** [comment]
It looks like it built:

```
[ 29%] Building C object AmberTools/src/nab/semantics/CMakeFiles/rule_parse.dir/parse2.c.o
[ 29%] Linking C executable rule_parse
[ 29%] Built target rule_parse
```

**Comment by mikemhenry** [comment]
@jaimergp Yah I will look back into where it puts it/where it looks for it. I would have expected some binary error about not being able to run an x86 binary on arm, so it might be looking in the wrong spot, good idea.

**Comment by jchodera** [comment]
@mikemhenry @jaimergp @simonbray @swails Any idea if this PR to add support for `arm` M1/M2 macs is ready to go?

**Comment by jchodera** [comment]
@swails: Any chance you (or someone you might know in the Amber community) would have the ability to help with this?

**Comment by jchodera** [comment]
@mikemhenry : Would be great if you could give this a try!

**Comment by mikemhenry** [comment]
@jaimergp thanks for the pointers
@jchodera I will put this on my project board

**Comment by mikemhenry** [comment]
working on this now...

**Comment by mikemhenry** [comment]
@jaimergp I'm getting the same error :upside_down_face: 


```
[  7%] Parsing nab rules and generating checkexpr.c
/bin/sh: rule_parse: command not found
```

I'm pretty sure I did things correctly RE build host tools, then point to them when building the source.
Looking at this: https://ambermd.org/pmwiki/pmwiki.php/Main/CMake-Cross-Compiling
It says it should build `rule_parse` but confusingly the build report doesn't list it:

```
  Building Tools: 

-- leap nab ucpp-1.3 
--                            NOT Building Tools: 
-- ************************************************************************** 
```

So I'm not sure if the wiki is out of date or something, but it seems to really be refusing to build  `rule_parse`

File: recipe/build.sh [hunk 1]
```diff
@@ -52,18 +59,83 @@ if [[ "$target_platform" == osx* ]]; then
...
+EOF
 
-make && make install
+    patch cmake/CompilerFlags.cmake < remove_mtune_native_flag.patch
+
+fi
+
+if [[ "${build_platform}" != "${target_platform}" ]]; then
+    # Build host tools first
+    mkdir -p ${PREFIX}../host_tools
```
Comment (jaimergp) [inline recipe/build.sh:98]:
```suggestion
    mkdir -p ${BUILD_PREFIX}/amber_host_tools
```

These tools will be native to the running arch, so they should be in BUILD_PREFIX.

File: recipe/build.sh [hunk 2]
```diff
@@ -52,18 +59,83 @@ if [[ "$target_platform" == osx* ]]; then
...
+    mkdir -p build_host_tools
+    cd build_host_tools || exit
```
Comment (jaimergp) [inline recipe/build.sh:100]:
```suggestion
    cd build_host_tools
```

We in `set -e`. `exit` (with no code) will make this script report a successful run.

File: recipe/build.sh [hunk 3]
```diff
@@ -52,18 +59,83 @@ if [[ "$target_platform" == osx* ]]; then
...
+    cmake ${CMAKE_ARGS} ${SRC_DIR} ${CMAKE_FLAGS} \
+        -DBUILD_HOST_TOOLS=TRUE \
+        -DCOMPILER=MANUAL \
+        -DCMAKE_INSTALL_PREFIX=${PREFIX}../host_tools
```
Comment (jaimergp) [inline recipe/build.sh:104]:
```suggestion
        -DCMAKE_INSTALL_PREFIX="${BUILD_PREFIX}/amber_host_tools"
```

File: recipe/build.sh [hunk 4]
```diff
@@ -52,18 +59,83 @@ if [[ "$target_platform" == osx* ]]; then
...
+
+    make
+    make install
+    # Now build the package
+    mkdir -p build
+    cd build || exit
+    cmake ${CMAKE_ARGS} ${SRC_DIR} ${CMAKE_FLAGS} \
+        -DBUILD_HOST_TOOLS=FALSE \
+        -DUSE_HOST_TOOLS=TRUE \
+        -DHOST_TOOLS_DIR=${PREFIX}../host_tools \
```
Comment (jaimergp) [inline recipe/build.sh:114]:
```suggestion
        -DHOST_TOOLS_DIR="${BUILD_PREFIX}/amber_host_tools" \
```

File: recipe/build.sh [hunk 5]
```diff
@@ -52,18 +59,83 @@ if [[ "$target_platform" == osx* ]]; then
...
+        -DBUILD_GUI=${BUILD_GUI} \
+        -DCHECK_UPDATES=FALSE \
+        -DTRUST_SYSTEM_LIBS=TRUE
+
+    make
+    make install
+else
+    # Build AmberTools with cmake
+    mkdir -p build
+    cd build || exit
```
Comment (jaimergp) [inline recipe/build.sh:127]:
```suggestion
    cd build
```

File: recipe/meta.yaml [hunk 6]
```diff
@@ -38,8 +45,8 @@ requirements:
     - cmake         # [not osx]
     - cmake 3.20.*  # [osx]
     - perl
-    - llvm-openmp  # [osx]
-    - libgomp      # [linux and not aarch64]
+    - llvm-openmp  # [osx and build_platform != "linux-64"]
```
Comment (jaimergp) [inline recipe/meta.yaml:48]: Why is this needed?

**Comment by mikemhenry** [comment]
@jaimergp thanks so much for your suggestions! I've learned a lot :)

Looks like we are still getting the same error unfortunately

File: recipe/meta.yaml (hunk 6 above)
Comment (mikemhenry) [reply recipe/meta.yaml:48]: I'm not sure, I can remove it and see if it changes anything. That change came from here:
https://github.com/conda-forge/ambertools-feedstock/commit/e260850480f6c1ba0696c31a68ed94802715ad99

File: recipe/build.sh [hunk 7]
```diff
@@ -52,18 +59,83 @@ if [[ "$target_platform" == osx* ]]; then
...
+        -DCMAKE_INSTALL_PREFIX="${BUILD_PREFIX}/amber_host_tools"
+
+    make
+    make install
+    # Now build the package
+    mkdir -p build
+    cd build || exit
+    cmake ${CMAKE_ARGS} ${SRC_DIR} ${CMAKE_FLAGS} \
+        -DBUILD_HOST_TOOLS=FALSE \
+        -DUSE_HOST_TOOLS=TRUE \
```
Comment (richardjgowers) [inline recipe/build.sh:113]: @mikemhenry I think this is what is causing the errors.

There's a cmake file in `nab/semantics/` which at the top says `#This file builds rule_parse, a component of the NAB compiler.
#NOTE: it'snot included at all if USE_HOST_TOOLS is enabled.`.

Assuming that this `HOST_TOOLS` flag is doing useful work elsewhere, I think maybe patching this CMake to make `rules_parse` get built even if this is set to true might fix this?

Comment (richardjgowers) [reply recipe/build.sh:113]: So I think patching lines 165-167 of `Ambertools/src/nab/CMakeLists.txt` will let you build `rules_parse`

**Comment by mikemhenry** [comment]
Okay so it builds it, I just need to make sure it ends up in the path then.

**Comment by mikemhenry** [comment]
So the patch seems to work, but now I'm second guessing if we need it. Looking at an older build, it confusingly did build `rules_parse`
```
[ 23%] Building C object AmberTools/src/nab/semantics/CMakeFiles/rule_parse.dir/parse2.c.o
[ 26%] Linking C executable rule_parse
[ 26%] Built target rule_parse
[ 28%] [BISON][NabYacc] Building parser with bison 3.8
nabgrm.y: warning: 3 shift/reduce conflicts [-Wconflicts-sr]
nabgrm.y: note: rerun with option '-Wcounterexamples' to generate conflict counterexamples
[ 31%] Parsing nab rules and generating checkexpr.c
/bin/sh: rule_parse: command not found
```

However, looking at `AmberTools/src/nab/semantics/CMakeLists.txt 

```
#This file builds rule_parse, a component of the NAB compiler.
#NOTE: it'snot included at all if USE_HOST_TOOLS is enabled.

set(RULEPARSE_SOURCE parse2.c)

add_executable(rule_parse ${RULEPARSE_SOURCE})
set_property(TARGET rule_parse PROPERTY COMPILE_OPTIONS ${NO_OPT_CFLAGS})

#install rule_parse in host tool mode
if(BUILD_HOST_TOOLS)
    install(TARGETS rule_parse DESTINATION ${BINDIR})
endif()
```

which is what we are including, it looks like it should install `rule_parse` in ${BINDIR} but that doesn't seem to either be happening, or BINDIR isn't in our path

**Comment by ghutchis** [comment]
I'm guessing you have access to an M1 - otherwise I have a GitHub self-hosted runner setup for builds.

**Comment by mikemhenry** [comment]
> I'm guessing you have access to an M1 - otherwise I have a GitHub self-hosted runner setup for builds.

I do! But I am interested on how you set up the self-hosted runner. Are you using containers or some other virtualization solution?

**Comment by ghutchis** [comment]
For now, the runner is set as an unprivileged user. At some point, I'll set it up as a container, but it's reasonable for now.

**Comment by ghutchis** [comment]
Ping - happy to provide access to the GH runner if it would help with the migration. It mostly just compiles `avogadrolibs` builds and has plenty of capacity.

**Comment by mikemhenry** [comment]
Thanks but first I want to try and see if I can get the new ambertools 2023 build on osx arm since if that works, then we should be good.

**Comment by hadim** [comment]
I would also be interested in having ambertools for osx-arm64. How far are you to get this done?

**Comment by mikemhenry** [comment]
@hadim I will have an update today!

**Comment by mikemhenry** [comment]
Will be working on this PR for 2023 
https://github.com/conda-forge/ambertools-feedstock/pull/116
